### PR TITLE
Change edition on url control.

### LINKED ISSUE #501: Change edition on url change
When changing the edition in  your url, this edition should be loaded. I think this is not the case right now.

### PR DESCRIPTION
**Pull request by jitsedesmet**
This PR closes #501.

### CODE REVIEW

File: data-injector/src/api/calls/studentCalls.ts [hunk 1]
```diff
@@ -6,7 +6,7 @@ import {
     getEntityOnUrl,
     getQueryUrlFromParams,
 } from "./baseCalls";
-import { IStudent, OsocExpericience, Status, studentCollectionName } from "../entities/StudentEntity";
+import { IStudent, OscExperience, Status, studentCollectionName } from "../entities/StudentEntity";
```
Comment (andpuydt) [inline data-injector/src/api/calls/studentCalls.ts:9]: Why was this changed?

File: frontend/src/components/project/projectList.tsx [hunk 2]
```diff
@@ -3,22 +3,29 @@ import styles from "../../styles/projects/projectList.module.css";
...
     const router = useRouter();
-    const transformer = useEditionApplicationPathTransformer();
+    const query = router.query as { id?: string };
+    const selectedProjectId = query.id;
+    const routerAction = useRouterPush();
     const currentUserIsAdmin = useCurrentAdminUser();
 
-    let { data, error } = useSwrWithEdition(apiPaths.projectsByEdition, getAllProjectsFromPage);
-    data = data || [];
+    let { data: receivedProjects, error: projectsError } = useSwrWithEdition(
```
Comment (jitsedesmet) [inline frontend/src/components/project/projectList.tsx:21]: let

File: frontend/src/components/user/invitationButton.tsx [hunk 3]
```diff
@@ -48,7 +48,7 @@ export default function InvitationButton() {
             invitationToken: postedInvitation.token,
         });
 
-        const registrationUrl = applicationPaths.base + "/" + transformer(url);
+        const registrationUrl = applicationPaths.base + "/" + url;
```
Comment (jitsedesmet) [inline frontend/src/components/user/invitationButton.tsx:51]: Is the transformer still needed in this file?

File: frontend/src/hooks/routerHooks.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+/**
+ * Url used by NextRouter.
+ */
+export type Url = string | UrlObject;
+
+export type RouterAction = (url: Url, as?: string, options?: TransitionOptions) => Promise<boolean>;
+
+export function useRouterPush(): RouterAction {
```
Comment (jitsedesmet) [inline frontend/src/hooks/routerHooks.ts:41]: docs

File: frontend/src/hooks/routerHooks.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+                newUrl.query = { ...(url.query || {}), edition: edition?.name };
+                return router.push(newUrl, as, options);
+            }
+        } else {
+            return router.push(url, as, options);
+        }
+    };
+}
+
+export function useRouterReplace(): RouterAction {
```
Comment (jitsedesmet) [inline frontend/src/hooks/routerHooks.ts:61]: docs

File: frontend/src/components/util/routeInjector.tsx [hunk 6]
```diff
@@ -15,7 +15,8 @@ import { pathIsAuthException } from "../../utility/pathUtil";
 
 export default function RouteInjector({ children }: any) {
     const router = useRouter();
-    const replace = router.replace;
+    // We can not inject here, it will provide problems and race conditions.
```
Comment (andpuydt) [inline frontend/src/components/util/routeInjector.tsx:18]:
```suggestion
    // We cannot inject here, it will provide problems and race conditions.
```

File: frontend/src/components/util/routeInjector.tsx [hunk 7]
```diff
@@ -45,10 +46,15 @@ export default function RouteInjector({ children }: any) {
...
     );
-    const latestEdition = availableEditions?.at(availableEditions?.length - 1);
+    const activeAvailableEditions = availableEditions
+        ? availableEditions.filter((edition) => edition.active)
+        : undefined;
+    const latestEdition = activeAvailableEditions?.at(activeAvailableEditions?.length - 1);
     const latestEditionName = latestEdition?.name;
 
     useEffect(() => {
+        // Some parts of the code are commented out because we think it'll help understand the choices that were made.
```
Comment (andpuydt) [inline frontend/src/components/util/routeInjector.tsx:56]:
```suggestion
        // Some parts of the code are commented out because we think it'll help to understand the choices that were made.
```

File: frontend/src/hooks/utilHooks.ts [hunk 8]
```diff
@@ -41,12 +41,22 @@ export function useEditionAPIUrlTransformer(): (url: string) => string {
...
     const [editionUrl] = useEdition();
     const { data: edition } = useSWR(editionUrl, getEditionOnUrl);
-    return (url) => getQueryUrlFromParams(url, { edition: edition?.name });
+
+    return (url) => {
+        // make sure only a single edition is contained in the url
+        if (edition) {
+            if (url.includes("?edition=") || url.includes("&edition=")) {
+                // Edition already set -> remove
+                return url;
```
Comment (jitsedesmet) [inline frontend/src/hooks/utilHooks.ts:50]: hmmm.... that's not a remove? If i logout, what happens? If the path at logout is okey, this code can get rerolled

File: frontend/src/components/util/routeInjector.tsx [hunk 9]
```diff
@@ -57,18 +63,24 @@ export default function RouteInjector({ children }: any) {
...
+            routerAction({
                 query: { ...query, edition: contextEditionName },
             }).catch(console.log);
         }
+         */
 
         // You do not have a path edition set but your config edition is set and defined -> set path edition
-        if (contextEditionName && !routerEditionName) {
-            replace({
+        // -> Not allowed since the query start with and empty object, and you have no way as to know if you changed the edition.
```
Comment (andpuydt) [inline frontend/src/components/util/routeInjector.tsx:76]:
```suggestion
        // -> Not allowed since the query starts with an empty object and you have no way of knowing whether you changed the edition.
```

File: frontend/src/components/util/routeInjector.tsx [hunk 10]
```diff
@@ -57,18 +63,24 @@ export default function RouteInjector({ children }: any) {
...
+        /*
+        if (contextEditionName && !routerEditionName && false) {
+            routerAction({
                 query: { ...query, edition: contextEditionName },
             }).catch(console.log);
         }
+         */
 
         // You can not get the edition that is saved in the context -> Might be because it was removed.
```
Comment (andpuydt) [inline frontend/src/components/util/routeInjector.tsx:85]:
```suggestion
        // You cannot get the edition that is saved in the context -> Might be because it was removed.
```

File: frontend/src/context/globalContext.tsx [hunk 11]
```diff
@@ -13,6 +13,8 @@ type ProviderProps = Record<string, unknown>;
 interface GlobalContextProps {
     // Undefined is de default value, it means the url has not been initialised.
     // Null means the edition is know, it is null.
+    // We can not delete this, even-though we always have an edition in our url.
```
Comment (andpuydt) [inline frontend/src/context/globalContext.tsx:16]:
```suggestion
    // We cannot delete this, even-though we always have an edition in our url.
```

File: frontend/src/context/globalContext.tsx [hunk 12]
```diff
@@ -13,6 +13,8 @@ type ProviderProps = Record<string, unknown>;
 interface GlobalContextProps {
     // Undefined is de default value, it means the url has not been initialised.
```
Comment (andpuydt) [inline frontend/src/context/globalContext.tsx:14]:
```suggestion
    // Undefined is the default value, it means the url has not been initialised.
```

File: frontend/src/hooks/utilHooks.ts (hunk 8 above)
Comment (andpuydt) [reply frontend/src/hooks/utilHooks.ts:50]: The edition stays in the url on logout

**Review by andpuydt — APPROVED** [review]

**Review by andpuydt — APPROVED** [review]
